middle_end/flambda/from_lambda/cps_conversion.ml

```diff
@@ -453,181 +691,166 @@ let rec cps_non_tail env (lam : L.lambda) (k : Ident.t -> Ilambda.t)
+              ~is_exn_handler:false
+              ~params:[body_result, Not_user_visible, Pgenval]
+              ~recursive:Nonrecursive
+              ~body:(fun env poptrap_continuation ->
+                let_cont_with_extra_params env ~continuation_on_stack:false
+                  ~is_exn_handler:false
+                  ~params:[]
+                  ~recursive:Nonrecursive
+                  ~body:(fun env body_continuation ->
+                    apply_cont_with_extra_args env
+                      body_continuation
+                      (Some (I.Push {
                         exn_handler = handler_continuation;
-                      }),
-                      []);
-                  handler = body;
-                };
-              handler = Apply_cont (after_continuation,
-                Some (I.Pop { exn_handler = handler_continuation; }),
-                [Ilambda.Var body_result]);
-            };
-          handler;
-        };
-      handler = k result_var;
-    }
+                      }))
+                      [])
+                  ~handler:(fun env ->
+                    cps_tail env body poptrap_continuation
+                      handler_continuation))
+              ~handler:(fun env ->
+                apply_cont_with_extra_args env
+                  after_continuation
+                  (Some (I.Pop { exn_handler = handler_continuation; }))
+                  [Ilambda.Var body_result]))
+          ~handler:(fun env ->
+            cps_tail env handler after_continuation k_exn))
+      ~handler:(fun env -> k env result_var)
   | Lifthenelse (cond, ifso, ifnot) ->
     let lam = switch_for_if_then_else ~cond ~ifso ~ifnot in
     cps_non_tail env lam k k_exn
   | Lsequence (lam1, lam2) ->
     let ident = Ident.create_local "sequence" in
     cps_non_tail env (L.Llet (Strict, Pgenval, ident, lam1, lam2)) k k_exn
   | Lwhile (cond, body) ->
-    let loop = rec_catch_for_while_loop cond body in
+    let env, loop = rec_catch_for_while_loop env cond body in
     cps_non_tail env loop k k_exn
   | Lfor (ident, start, stop, dir, body) ->
-    let loop = rec_catch_for_for_loop ident start stop dir body in
+    let env, loop = rec_catch_for_for_loop env ident start stop dir body in
     cps_non_tail env loop k k_exn
   | Lassign (being_assigned, new_value) ->
     if not (Env.is_mutable env being_assigned) then begin
       Misc.fatal_errorf "Lassign on non-mutable variable %a"
         Ident.print being_assigned
     end;
-    cps_non_tail_simple env new_value (fun new_value ->
-        name_then_cps_non_tail "assign"
-          (I.Assign { being_assigned; new_value; })
-          k k_exn)
+    cps_non_tail_simple env new_value (fun env new_value ->
+        let env, new_id, new_kind = Env.new_id_for_mutable env being_assigned in
+        I.Let (new_id, User_visible, new_kind,
+          Simple new_value,
+          name_then_cps_non_tail env "assign"
+            (I.Simple (Const (Const_base (Const_int 0))))
```

Comment:
You could use `I.Simple Lambda.const_unit`, which is both more concise and more explicit.